packages/modules/e3dc/device.py

```diff
@@ -0,0 +1,181 @@
+    component_config = E3dcInverterSetup(configuration=E3dcInverterConfiguration())
+    component_config.id = num
+    run_device_legacy(create_legacy_device_config(address1,
+                                                  address2,
+                                                  read_ext,
+                                                  pvmodul,
+                                                  num), component_config)
+
+
+def read_legacy_batv19(address1: str,
+                       address2: str, read_ext: int,
+                       pvmodul: str,
+                       num: int) -> None:
+    # für openwbv19 können wmit der bisherigen parametrisierung zwei ip_addressen
+    # ausgelesen werden
+    # ebenso wird bei Speicheraufruf beides (Speicher und PV ausgelesen
+    # in openwb v2.0 geht nur noch eine IP adresse und die Pv muss
+    # separate ausgelesen werden
+    addresses = [address for address in [address1, address2] if address != "none"]
+    log.debug('e3dc IP-Adresse1: ' + address1)
+    log.debug('e3dc IP-Adresse2: ' + address2)
+    log.debug('e3dc read_ext: ' + str(read_ext))
+    log.debug('e3dc pvmodul: ' + pvmodul)
+    log.debug('e3dc id: ' + str(num))
+    soc = 0
+    count = 0
+    power = 0
+    pv_external = 0
+    pv = 0
+    pv_other = pvmodul != "none"
+    for address in addresses:
+        log.debug("Ip: %s, read_external %d pv_other %s", address, read_ext, pv_other)
+        count += 1
+        with modbus.ModbusTcpClient_(address, port=502) as client:
+            socwk, powerwk = read_bat(client)
+            soc += socwk
+            power += powerwk
+            pvwk, pv_externalwk = read_inverter(client, read_ext)
+            pv += pvwk
+            pv_external += pv_externalwk
+    soc = soc / count
+    log.debug("Soc %d power %d pv %d pv_external %d count ip %d",
+              soc, power, pv, pv_external, count)
+    counter_import, counter_export = sim_count(power, prefix="speicher")
+    get_bat_value_store(1).set(BatState(power=power, soc=soc, imported=counter_import, exported=counter_export))
+    # pv_other sagt aus, ob WR definiert ist, und dessen PV Leistung auch gilt
+    # wenn 0 gilt nur PV und pv_external aus e3dc
+    pv_total = pv + pv_external
+    # Wenn wr1 nicht definiert ist, gilt nur die PV Leistung die hier im Modul ermittelt wurde
+    # als gesamte PV Leistung für wr1
+    if not pv_other or pv_total != 0:
+        # Wenn wr1 definiert ist, gilt die bestehende PV Leistung aus Wr1 und das was hier im Modul ermittelt wurde
+        # als gesamte PV Leistung für wr1
+        if pv_other:
+            try:
+                pv_total = pv_total + files.pv[0].power.read()
+            except Exception:
+                pass
+        log.debug("wr update pv_other %s pv_total %d", pv_other, pv_total)
+        _, counter_pv = sim_count(pv_total, prefix="pv")
+        get_inverter_value_store(1).set(InverterState(exported=counter_pv, power=pv_total))
```

Comment:
Mein Verständnis sagt, hier wird beim ersten Durchlauf PV-Leistung abgespeichert. Beim nächsten Durchlauf wird die Leistung aus dem vorherigen Durchlauf (abgerufen durch `files.pv[0].power.read()`) zu der aktuellen Leistung hinzuaddiert. Die Leistung müsste ins unermessliche ansteigen. Oder habe ich das ganz falsch verstanden?

Reply:
Siehe Erklärung oben (Fall c) .

Reply:
Verstanden... Glaube ich. Habe dann ein paar andere Fragen:

1. Wenn `pv_total != 0`, dann führt es SimCount aus und speichert den `InverterState`. Aber was passiert mit dem letzten Wert des Tages? Sagen wir mal die PV-Anlage meldet als letzten Wert des Tages 50 Watt. Kurz danach kommt 0 Watt. Dann würden die 50 Watt die ganze Nacht durch stehen bleiben, oder? Wenn am nächsten Tag die Anlage wieder aufwacht und sagen wir mal der erste Wert des Tages ist 10 Watt, dann würde SimCount den neuen Zählerstand so errechnen, als würden die Nacht durch im Schnitt 30 Watt produziert worden und der Zähler geht ungewollt hoch. Wenn du den Fall mit der SolarEdge-PV hast, dann würden in der Nacht plötzlich ändere Zählerstände da stehen, denn dann bleiben die Werte aus den SolarEdge-Registern stehen und werden nicht überschrieben. Ich meine die Bedingung in Zeile 161 müsste ersatzlos gestrichen.
2. Zu dem
    ```python
    except Exception:
        pass
    ```
    Mir fallen zwei mögliche Exceptions ein: `FileNotFoundError`, wenn die Datei nicht gefunden wurde und `ValueError`, wenn die Zahl in der Datei kein gültiger Integer ist. Ich meine das sollte beides nicht vorkommen, das würde für einen tiefergehenden Fehler sprechen. Wenn ich da nichts übersehe würde ich vorschlagen, den Fehler nicht zu fangen. In dem Fall läuft die Exception hoch und wir am Ende geloggt und gibt damit dem Nutzer die Möglichkeit den Fehler zu finden. Wenn du es ganz genau machen willst könntest du den read_legacy code noch wrappen in:
    ```python
    with SingleComponentUpdateContext(ComponentInfo.from_component_config(E3dcBatSetup(id=num))):
        do_stuff()
    ```
    Dann würden Fehler in `do_stuff()` auf der Statusseite angezeigt.

Reply:
Zu 1)
  if not pv_other or pv_total != 0:
Pv_other ist True wenn eine pv Anlage definiert ist. 
Also wenn pv nicht definiert ist, oder eine pv Leistung vom e3dc  gemessen wurde, dann wird die Pv Leistung geschrieben. Hast recht ist heute wohl überflüssig. Ich nehme es raus und teste es,
Zu 2)
Hm... war wohl drin wenn um zu checken wenn die ramdisk noch nicht ready war.
Hast recht ist heute wohl auch überflüssig. Ich nehme es raus und teste es,